azure-iot-device/azure/iot/device/provisioning/internal/request_response_provider.py

```diff
@@ -0,0 +1,80 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import json
+from ..models.registration_result import RegistrationResult, RegistrationState
+
+logger = logging.getLogger(__name__)
+
+
+class RequestResponseMachine(object):
+    def __init__(self, state_based_provider):
+
+        self._state_based_provider = state_based_provider
+        self._state_based_provider.on_state_based_provider_message_received = self._receive_response
+
+        self.on_response_received = None
+
+    def send_request(self, publish_action):
```

Comment:
>send_request [](start = 8, length = 12)

I was thinking that the utility of this class would be that send_request would take a callback along with the request. It would publish the request, and not call the callback until it receives a response with the matching $rid value.  This way the layer above wouldn't have to add special logic to wait for the response because the waiting would be handled inside the RequestResponseMachine.